atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java

```diff
@@ -69,157 +74,191 @@
+    public static ProfilingKeyValueService create(KeyValueService delegate, long slowLogThresholdMillis) {
+        return new ProfilingKeyValueService(delegate, slowLogThresholdMillis);
     }
 
-    private static void logCellsAndSize(String method, String tableName, int numCells, long sizeInBytes, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} on table {} for {} cells of overall size {} bytes took {} ms.",
-                method, tableName, numCells, sizeInBytes, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    @FunctionalInterface
+    interface LoggingFunction {
+        void log(String fmt, Object... args);
     }
 
-    private static void logTime(String method, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} took {} ms.", method, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    @FunctionalInterface
+    interface TriConsumer<A,B,C> {
+        void accept(A aa, B bb, C cc);
     }
 
-    private static void logTimeAndTable(String method, String tableName, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} on table {} took {} ms.",
+    private static void logCellsAndSize(LoggingFunction logger, String method, String tableName, int numCells, long sizeInBytes, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} for {} cells of overall size {} bytes took {} ms.",
+                tableName, numCells, sizeInBytes, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    }
+
+    private static void logTime(LoggingFunction logger, String method, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} took {} ms.", method, stopwatch.elapsed(TimeUnit.MILLISECONDS));
+    }
+
+    private static void logTimeAndTable(LoggingFunction logger, String method, String tableName, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} took {} ms.",
                 method, tableName, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private static void logTimeAndTableCount(String method, int tableCount, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} for {} tables took {} ms.",
+    private static void logTimeAndTableCount(LoggingFunction logger, String method, int tableCount, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} for {} tables took {} ms.",
                 method, tableCount, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
-    private static void logTimeAndTableRange(String method, String tableName, RangeRequest range, Stopwatch stopwatch) {
-        log.trace("Call to KVS.{} on table {} with range {} took {} ms.",
+    private static void logTimeAndTableRange(LoggingFunction logger, String method, String tableName, RangeRequest range, Stopwatch stopwatch) {
+        logger.log("Call to KVS.{} on table {} with range {} took {} ms.",
                 method, tableName, range, stopwatch.elapsed(TimeUnit.MILLISECONDS));
     }
 
+    private void maybeLog(Runnable runnable, Predicate<Stopwatch> shouldSlowLog, BiConsumer<LoggingFunction, Stopwatch> logger) {
```

Comment:
I like this injected structure, but there is still quite a bit of duplicate logic.  We should be able to isolate most of the logic to the most complex of the impls (the last one) and impl the rest using it.